### PR TITLE
[ticket/13779] Set new auth options to the role only if matching the role type

### PR DESCRIPTION
**Pull request by rxu**
Migrations' permission tool allows setting permissions to the role which
doesn't match the role type, e.g. m_ permissions for u_ role types and so on.
As one of side effects, this may lead to granting users moderative/admin
permissions silently.
With this patch the only new permissions matching the role type will be set.

<a href="https://tracker.phpbb.com/browse/PHPBB3-13779">PHPBB3-13779</a>.

### CODE REVIEW

File: phpBB/phpbb/db/migration/tool/permission.php [hunk 1]
```diff
@@ -429,9 +429,23 @@ public function permission_set($name, $auth_option, $type = 'role', $has_permiss
 						FROM ' . ACL_ROLES_TABLE . '
 						WHERE role_id = ' . $role_id;
 					$this->db->sql_query($sql);
-					$role_name = $this->db->sql_fetchfield('role_name');
-
-					return $this->permission_set($role_name, $auth_option, 'role', $has_permission);
+					$role_data = $this->db->sql_fetchrow();
+					$role_name = $role_data['role_name'];
+					$role_type = $role_data['role_type'];
```
Comment (VSEphpbb) [inline phpBB/phpbb/db/migration/tool/permission.php]: role_type is not retrieved in the query

Comment (rxu) [inline phpBB/phpbb/db/migration/tool/permission.php]: Fixed, thanks a lot.

File: phpBB/phpbb/db/migration/tool/permission.php [hunk 2]
```diff
@@ -425,13 +425,27 @@ public function permission_set($name, $auth_option, $type = 'role', $has_permiss
...
+
+					// Filter new auth options to match the role type: a_ | f_| m_ | u_
+					// Set new auth options to the role only if options matching the role type were found
+					$auth_option = array_filter($auth_option,
+						function ($option) use ($role_type)
+						{
+							return strpos($auth_option, $auth_type) === 0;
```
Comment (VSEphpbb) [inline phpBB/phpbb/db/migration/tool/permission.php]: $auth_option should be $option? And $auth_type should be $role_type?

Comment (rxu) [inline phpBB/phpbb/db/migration/tool/permission.php]: Hmm yeah, but ```$auth_option``` is correct as it's the method's input options array.

Comment (VSEphpbb) [inline phpBB/phpbb/db/migration/tool/permission.php]: But you're passing the $auth_option array values it in to the closure as `$option` which is what is being tested in strpos...no?

File: phpBB/phpbb/db/migration/tool/permission.php [hunk 3]
```diff
@@ -425,13 +425,27 @@ public function permission_set($name, $auth_option, $type = 'role', $has_permiss
...
 						WHERE role_id = ' . $role_id;
 					$this->db->sql_query($sql);
-					$role_name = $this->db->sql_fetchfield('role_name');
-
-					return $this->permission_set($role_name, $auth_option, 'role', $has_permission);
+					$role_data = $this->db->sql_fetchrow();
+					$role_name = $role_data['role_name'];
+					$role_type = $role_data['role_type'];
+
+					// Filter new auth options to match the role type: a_ | f_| m_ | u_
```
Comment (VSEphpbb) [inline phpBB/phpbb/db/migration/tool/permission.php]: maybe clean up the comment ... space between `f_|` :smile:

**Comment by Nicofuma** [comment]
I'd like to be able to warn the developer when it happens but for now I don't see how...

**Comment by nickvergessen** [comment]
@rxu can we somehow add unit tests for this?
https://github.com/phpbb/phpbb/blob/3.1.x/tests/dbal/migrator_tool_permission_test.php

**Comment by rxu** [comment]
@nickvergessen I'll look into possible tests cases.

**Comment by rxu** [comment]
It looks like assigning permissions fail anyway for the group having more than 1 role assigned as the function only selects 1 role per group.
For the ```'role'``` mode, it can grant permissions regardless of the role type (f.e. ```a_``` permission for the ```u_``` role type, is it intended?).

**Comment by rxu** [comment]
@nickvergessen Tests added.

**Comment by Nicofuma** [comment]
:+1:

**Pull request merged**